content/blog/2023-12-11-A DSL for Metastability.md

```diff
@@ -0,0 +1,237 @@
+That is, the agent receives an input state, goes through all of the actions in its code via a series of small steps, and returns another state.
+From the point of view of an external observer, all of this is just a big step.
+Most of the actions in the grammar are self-explanatory.
+Here are some details:
+
+1. Predicates have no effect on the state. Only their value at the time is important. Therefore, all an interpreter has to do is to evaluate the predicate according to some simple rules, and to take the appropriate action based on the value.
+2. As I also mentioned in the syntax section, the Kleene star is always guarded by a predicate for now. Therefore, the action specified by the star operation keeps being applied to the state in small-step fashion until the value of the predicate becomes $\texttt{false}$.
+3. When multiple agents are applying their code to the global state, their actions commute since they are working on different parts of the global state. The concurrency related to messaging is handled in the concurrent execution semantics.
+
+### Concurrent Execution Semantics
+Once we have a number of agents and know how to interpret their code, it is time to execute the entire system.
+I have assumed that the system is being executed in discrete time with its units called *steps*.
+There is an entity called the "network", that at each step receives the requests sent by the agents, and delivers the requests that are due for reception.
+The network can be modeled as synchronous or asynchronous, and I have opted for synchronous for now.
+The concurrency is handled by the order in which the network decides to receive requests from the agents and to deliver requests to the agents.
+It can reorder the requests however it wants, subject to potential constraints (e.g., FIFO between the same sender and receiver).
+
+Note that there are entire languages and logics just to study this networking behavior.
+This is, of course, bread and butter for distributed systems, but it is still subject to on-going research.
+I have not tried to express these networking behaviors via existing frameworks, and I am planning to capture them with adequate infrastructure and API in the execution engine for the whole system.
+
+There is also an entity called the environment, which is responsible for *triggers*.
+The environment is not expressed by the DSL, and is assumed to operate based on pure choice: each "action" for the environment is just an arbitrary state transformation subject to the limitations imposed on the environment.
+These limitations are expressed by specifying the variables that the environment can touch.
+For instance, the code for some agent can be wrapped with a giant branching action with a guard $\texttt{not crashed}$, and a $\texttt{nop}$ for the else clause.
+The environment can then choose to crash and uncrash the agent however it likes.
```

Comment:
This conception of the environment makes sense to me, but I don't quite understand what a "trigger" is… does that terminology matter?

Reply:
That's a good point! To explain this I would have to unpack our definition of metastability, which is based on temporal logic, and it would take a lot of space. That is why I decided to leave it out. In summary, there exists a predicate P over environment actions, and a trigger is an initial period of the environment going crazy, until it "eventually" satisfies "always" P.

Reply:
Sounds great. Maybe it be worth just briefly mentioning that you are also interested in defining metastability formally so you can build an automatic detector, but not going into detail about what that definition actually is.